File: app/src/main/java/com/panda/ap_anime/anmi_details/DetailsActivity.java
package com.panda.ap_anime.anmi_details;

import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.widget.Toast;
import butterknife.BindView;
import butterknife.ButterKnife;
import carbon.widget.ImageView;
import carbon.widget.RecyclerView;
import carbon.widget.TextView;
import com.bumptech.glide.Glide;
import com.google.android.youtube.player.YouTubeBaseActivity;
import com.panda.ap_anime.Constants;
import com.panda.ap_anime.R;
import com.panda.ap_anime.anmi_details.details_repo.*;
import com.panda.ap_anime.c_anmiations_utils.CAnmiation;
import qenawi.panda.a_predator.network_Handeler.A_Predator_NWM;
import qenawi.panda.a_predator.network_Handeler.A_Predator_NetWorkManger;
import qenawi.panda.a_predator.network_Handeler.A_Predator_Throwable;
import qenawi.panda.a_predator.network_Handeler.CService_DBase;

import java.util.ArrayList;
import java.util.HashMap;

import static com.panda.ap_anime.descover_fragment.DescoverFragment.ApiKey;

public class DetailsActivity extends YouTubeBaseActivity {
    CAnmiation cAnmiation;
    A_Predator_NetWorkManger a_predator_netWorkManger;
    @BindView(R.id.imageView)
    ImageView mainImageView;
    @BindView(R.id.descreption)
    TextView descreption;
    @BindView(R.id.recycler_categ)
    RecyclerView category_rv;
    @BindView(R.id.tralers_recycler)
    RecyclerView tralers_recycler;
    @BindView(R.id.stream_links)
    RecyclerView stream_links;
    private GenreAdapter genreAdapter;
    private TralersAdapter tralersAdapter;
    private StreamAdapter streamAdapter;
    int Id;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.anmi_details_activity);
        ButterKnife.bind(this);
        if (getIntent().getExtras() != null && getIntent().getExtras().containsKey("AnimeKey")) {
            Id = getIntent().getExtras().getInt("AnimeKey");
        } else return;
        initView();
        getData();
    }

    private void getData() {
        HashMap<String, String> Headrs = new HashMap<String, String>();
        HashMap<String, Object> Body = new HashMap<String, Object>() {
            {
                put("api_key", ApiKey);
                put("language", "en");

            }
        };

        // https://api.themoviedb.org/3/movie/166428/videos?api_key=7d1dd3f8f443ecde27c323485a170bdc&language=en-US


        String Url = "https://api.themoviedb.org/3/movie/" + Id;
        String UrlVedios = Url + "/videos";
        a_predator_netWorkManger.FetchData(new Anmi_Details_Response(), Headrs, Url, Body, new A_Predator_NWM.RequistResuiltCallBack() {
            @Override
            public <T extends CService_DBase> void Sucess(T data) {
                Anmi_Details_Response anmi_details_response = (Anmi_Details_Response) data;
                Glide.with(DetailsActivity.this).load(Constants.Movies.getImageBaseUrl("w500") + (anmi_details_response.getBackdropPath())).into(mainImageView);
                descreption.setText(anmi_details_response.getOverview());
                if (((Anmi_Details_Response) data).getGenres() != null && !((Anmi_Details_Response) data).getGenres().isEmpty())
                    genreAdapter.replaceData(((Anmi_Details_Response) data).getGenres());
            }

            @Override
            public void Faild(A_Predator_Throwable error) {
                if (error.GetDoAction())
                    Toast.makeText(DetailsActivity.this, error.getACtion(), Toast.LENGTH_SHORT).show();
                finish();
            }
        });

        a_predator_netWorkManger.FetchData(new TralerResponse(), Headrs, UrlVedios, Body, new A_Predator_NWM.RequistResuiltCallBack() {
            @Override
            public <T extends CService_DBase> void Sucess(T Resposne) {
                TralerResponse tralerResponse = (TralerResponse) Resposne;
                tralersAdapter.replaceData(tralerResponse.getResults());
            }

            @Override
            public void Faild(A_Predator_Throwable error) {
                if (error.GetDoAction())
                    Toast.makeText(DetailsActivity.this, error.getACtion(), Toast.LENGTH_SHORT).show();
            }
        });


    }

    private void initView() {
        cAnmiation = new CAnmiation(this, 150);
        a_predator_netWorkManger = new A_Predator_NetWorkManger(this, new A_Predator_NWM.BaseActionHandeler() {
            @Override
            public void ShowSnackBar(String t) {

            }

            @Override
            public void HandelUnAuthAction() {

            }

            @Override
            public void DetacHHandelers() {

            }
        });
        genreAdapter = new GenreAdapter(new ArrayList<Genre>(), id ->
        {

        });
        tralersAdapter = new TralersAdapter(new ArrayList<Trailer_obj>(), id -> {

        });
        streamAdapter = new StreamAdapter(new ArrayList<StreamObject>() {
            {
                add(new StreamObject("CimaClup", "720P", "sTeam0"));
                add(new StreamObject("KissCartoon", "720P", "sTeam0"));
                add(new StreamObject("AnimeLeek", "720P", "sTeam0"));
                add(new StreamObject("YouTube", "720P", "sTeam0"));

            }
        }, id ->
        {

        });

        stream_links.setLayoutManager(new LinearLayoutManager(this));
        stream_links.setHasFixedSize(true);
        stream_links.setAdapter(streamAdapter);


        category_rv.setLayoutManager(new GridLayoutManager(this, 4));
        category_rv.setHasFixedSize(true);
        category_rv.setAdapter(genreAdapter);

        tralers_recycler.setLayoutManager(new LinearLayoutManager(DetailsActivity.this, LinearLayoutManager.HORIZONTAL, false));
        tralers_recycler.setHasFixedSize(true);
        tralers_recycler.setAdapter(tralersAdapter);
    }

}

File: app/src/main/java/com/panda/ap_anime/anmi_details/DetailsActivity.kt
package com.panda.ap_anime.anmi_details

import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.widget.Toast
import com.bumptech.glide.Glide
import com.google.android.youtube.player.YouTubeBaseActivity
import com.panda.ap_anime.Constants
import com.panda.ap_anime.R
import com.panda.ap_anime.anmi_details.details_repo.*
import com.panda.ap_anime.c_anmiations_utils.CAnmiation
import kotlinx.android.synthetic.main.anmi_details_activity.*
import qenawi.panda.a_predator.network_Handeler.A_Predator_NWM
import qenawi.panda.a_predator.network_Handeler.A_Predator_NetWorkManger
import qenawi.panda.a_predator.network_Handeler.A_Predator_Throwable
import qenawi.panda.a_predator.network_Handeler.CService_DBase
import java.util.ArrayList
import java.util.HashMap
import android.content.Intent
import android.net.Uri
import android.view.KeyEvent
import okhttp3.Dispatcher

class DetailsActivity : YouTubeBaseActivity() {
    lateinit var cAnmiation: CAnmiation
    lateinit var a_predator_netWorkManger: A_Predator_NetWorkManger
    private var genreAdapter: GenreAdapter? = null
    private var tralersAdapter: TralersAdapter? = null
    private var streamAdapter: StreamAdapter? = null
    internal var Id: Int = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.anmi_details_activity)
        if (intent.extras != null && intent.extras!!.containsKey("AnimeKey")) {
            Id = intent.extras!!.getInt("AnimeKey")
        } else
            return
        initView()
        getData()
    }

    private fun getData() {
        val Headrs = HashMap<String, String>()
        val Body = object : HashMap<String, Any>() {
            init {
                put("api_key", Constants.TMDP_API_KEY)
                put("language", "en")

            }
        }

        // https://api.themoviedb.org/3/movie/166428/videos?api_key=7d1dd3f8f443ecde27c323485a170bdc&language=en-US


        val Url = "https://api.themoviedb.org/3/movie/$Id"
        val UrlVedios = "$Url/videos"
        a_predator_netWorkManger.FetchData(
            Anmi_Details_Response(),
            Headrs,
            Url,
            Body,
            object : A_Predator_NWM.RequistResuiltCallBack {
                override fun <T : CService_DBase> Sucess(data: T) {
                    val anmi_details_response = data as Anmi_Details_Response
                    Glide.with(this@DetailsActivity)
                        .load(Constants.Movies.getImageBaseUrl("w500") + anmi_details_response.backdropPath)
                        .into(imageView)
                    descreption.text = anmi_details_response.overview
                    if ((data as Anmi_Details_Response).genres != null && !(data as Anmi_Details_Response).genres!!.isEmpty())
                        genreAdapter!!.submitList((data as Anmi_Details_Response).genres)
                }

                override fun Faild(error: A_Predator_Throwable) {
                    if (error.GetDoAction())
                        Toast.makeText(this@DetailsActivity, error.aCtion, Toast.LENGTH_SHORT).show()
                    finish()
                }
            })

        a_predator_netWorkManger.FetchData(
            TralerResponse(),
            Headrs,
            UrlVedios,
            Body,
            object : A_Predator_NWM.RequistResuiltCallBack {
                override fun <T : CService_DBase> Sucess(Resposne: T) {
                    val tralerResponse = Resposne as TralerResponse
                    tralersAdapter!!.submitList(tralerResponse.results)
                }

                override fun Faild(error: A_Predator_Throwable) {
                    if (error.GetDoAction())
                        Toast.makeText(this@DetailsActivity, error.aCtion, Toast.LENGTH_SHORT).show()
                }
            })


    }

    private fun initView() {
        cAnmiation = CAnmiation(this, 150)
        a_predator_netWorkManger = A_Predator_NetWorkManger(this, object : A_Predator_NWM.BaseActionHandeler {
            override fun ShowSnackBar(t: String) {

            }

            override fun HandelUnAuthAction() {

            }

            override fun DetacHHandelers() {

            }
        })
        genreAdapter = GenreAdapter()

        tralersAdapter = TralersAdapter(object : TralersAdapter.OnSrelection {
            override fun selection(str: String) {

                val videoId = str
                val intent = Intent(Intent.ACTION_VIEW, Uri.parse("vnd.youtube:$videoId"))
                intent.putExtra("VIDEO_ID", videoId)
                startActivity(intent)
            }
        })

        val nn: ArrayList<StreamObject> = ArrayList()
        nn.add(StreamObject("CimaClup", "720P", "sTeam0"))
        nn.add(StreamObject("KissCartoon", "720P", "sTeam0"))
        nn.add(StreamObject("AnimeLeek", "720P", "sTeam0"))
        nn.add(StreamObject("YouTube", "720P", "sTeam0"))
        streamAdapter = StreamAdapter()
        streamAdapter?.submitList(nn)

        stream_links.layoutManager = LinearLayoutManager(this)
        stream_links.adapter = streamAdapter


        recycler_categ.layoutManager = LinearLayoutManager(this@DetailsActivity, LinearLayoutManager.HORIZONTAL, false)
        recycler_categ.adapter = genreAdapter

        tralers_recycler.layoutManager =
            LinearLayoutManager(this@DetailsActivity, LinearLayoutManager.HORIZONTAL, false)
        tralers_recycler.adapter = tralersAdapter
    }

}
